### PR TITLE
POC partition cleanup

### PR DESCRIPTION
**Pull request by chrismeyersfsu**
* For each job type, get all event partitions from the DB.
* Exclude partitions that are > the cutoff.
* Exclude partitions that have events for any jobs that are not
finished.
* Handle special case <tbl_name>_old_events

TODO:
* remove debug messages
* better debug messages
* remove minutely logic
* Wrap everything in a giant transaction !!!

### CODE REVIEW

File: awx/main/management/commands/cleanup_jobs.py [hunk 1]
```diff
@@ -21,7 +23,160 @@
...
+    event_tbl_name = unified_job_class_to_event_table_name(job_class)
+    event_tbl_name += f'_{suffix}'
+    return event_tbl_name
+
+
+def partition_suffix(d, minutely=False):
+    if not minutely:
+        return d.replace(microsecond=0, second=0, minute=0).strftime('%Y%m%d_%H')
+    else:
+        return d.replace(microsecond=0, second=0).strftime('%Y%m%d_%H')
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:48]: If `minutely` is True, seems like we would want the string to include the minute as well.

File: awx/main/management/commands/cleanup_jobs.py [hunk 2]
```diff
@@ -21,7 +23,160 @@
...
+        main_inventoryupdateevent_20210318_03
+    '''
+    if 'old_events' in part_name:
+        return None
+    p = re.compile('([a-z]+)_([a-z]+)_([0-9]+)_([0-9][0-9])')
+    m = p.match(part_name)
+    if not m:
+        return m
+    dt_str = f"{m.group(3)}_{m.group(4)}"
+    dt = datetime.datetime.strptime(dt_str, '%Y%m%d_%H').replace(tzinfo=pytz.UTC)
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:65]: Just a note in passing, timezones in general are something I'm a little worried about with partitioning. I thought I was setting the timezone based on a tower setting, but I can't find where I'm doing that now. Anyways, just wanted to highlight a potential gremlin.

File: awx/main/management/commands/cleanup_jobs.py [hunk 3]
```diff
@@ -21,7 +23,160 @@
...
+
+    def find_jobs_to_delete(self):
+        self.job_qs = self.job_class.objects.filter(created__lt=self.cutoff).values_list('pk', 'status', 'created')
+        for pk, status, created in self.job_qs:
+            if status not in ['pending', 'waiting', 'running']:
+                self.jobs_to_delete_count += 1
+                self.jobs_pk_list.append(pk)
+        self.jobs_no_delete_count = (
+            self.job_class.objects.filter(created__gte=self.cutoff) |
+            self.job_class.objects.filter(status__in=['pending', 'waiting', 'running'])
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:98]: I'm not sure how expensive this query is, but if it seems worth the optimization, could drop this and just keep track of the count in the for-loop that's just above this.

```diff
+ no_delete_count = 0
if status not in ['pending', 'waiting', 'running']:
                self.jobs_to_delete_count += 1
                self.jobs_pk_list.append(pk)
+ else:
+               no_delete_count += 1
```

Comment (jladdjr) [reply awx/main/management/commands/cleanup_jobs.py:98]: would still need `self.job_class.objects.filter(created__gte=self.cutoff)` of course

File: awx/main/management/commands/cleanup_jobs.py (hunk 1 above)
Comment (chrismeyersfsu) [reply awx/main/management/commands/cleanup_jobs.py:48]: yeah, sorry. This doesn't really support minutely. Need to delete that condition.

File: awx/main/management/commands/cleanup_jobs.py (hunk 3 above)
Comment (chrismeyersfsu) [reply awx/main/management/commands/cleanup_jobs.py:98]: > I'm not sure how expensive this query is, but if it seems worth the optimization, could drop this and just keep track of the count in the for-loop that's just above this.
> 
> ```diff
> if status not in ['pending', 'waiting', 'running']:
>                 self.jobs_to_delete_count += 1
>                 self.jobs_pk_list.append(pk)
> + else:
> +               self.jobs_no_delete_count += 1
> ```

The set of jobs being looped over are all less than or equal to `self.cutoff`. I think we need the entire query here.

File: awx/main/management/commands/cleanup_jobs.py [hunk 4]
```diff
@@ -21,7 +23,160 @@
...
+        ).count()
+
+    def identify_excluded_partitions(self):
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:101]: 👍

File: awx/main/management/commands/cleanup_jobs.py [hunk 5]
```diff
@@ -21,7 +23,160 @@
...
+    def delete_jobs(self):
+        if not self.dry_run:
+            self.job_class.objects.filter(pk__in=self.jobs_pk_list).delete()
+
+    def find_partitions_to_drop(self):
+        tbl_name = unified_job_class_to_event_table_name(self.job_class)
+        old_events_tbl_name = f"{tbl_name}_old_events"
+
+        with connection.cursor() as cursor:
+            cursor.execute(f"select to_regclass('{old_events_tbl_name}') is not null")
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:129]: > `to_regclass`

does this mean that `old_events_tbl_name` can sometimes represent a regex?

File: awx/main/management/commands/cleanup_jobs.py [hunk 6]
```diff
@@ -21,7 +23,160 @@
...
+            cursor.execute(query)
+            partitions_from_db = [r[0] for r in cursor.fetchall()]
+
+        partitions_dt = [partition_name_dt(p) for p in partitions_from_db if not None]
+        partitions_dt = [p for p in partitions_dt if not None]
+        partitions_dt.sort()
+
+        # Find partitions that match our drop date window
+        i = 0
+        for i, dt in enumerate(partitions_dt):
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:148]: 👍

File: awx/main/management/commands/cleanup_jobs.py [hunk 7]
```diff
@@ -21,7 +23,160 @@
...
+            if dt > self.cutoff:
+                break
+        partitions_dt = partitions_dt[0:i]
+
+        # convert datetime partition back to string partition
+        partitions_maybe_drop = set([dt_to_partition_name(dt) for dt in partitions_dt])
+
+        # Do not drop partition if there is a job that will not be deleted pointing at it
+        self.parts_to_drop = partitions_maybe_drop - self.parts_no_drop
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:157]: 👍

File: awx/main/management/commands/cleanup_jobs.py [hunk 8]
```diff
@@ -21,7 +23,160 @@
...
+
+        # Handle dropping special case <job_table_name>_old_events
+        epoch = get_event_partition_epoch()
+        if old_events_table_exists and epoch <= self.cutoff:
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:161]: 👍 

yep, no need to check for running jobs here, all records here happened prior to upgrade

File: awx/main/management/commands/cleanup_jobs.py [hunk 9]
```diff
@@ -21,7 +23,160 @@
...
+                if applied and created < applied:
+                    part_key = 'old_events'
+
+                part_drop.setdefault(part_key, True)
+
+        self.parts_no_drop = set([k for k, v in part_drop.items() if v is False])
+
+    def delete_jobs(self):
+        if not self.dry_run:
+            self.job_class.objects.filter(pk__in=self.jobs_pk_list).delete()
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:122]: 👍

Comment (jladdjr) [reply awx/main/management/commands/cleanup_jobs.py:122]:
```
else:
    <log what we would do in dry run?>
```

File: awx/main/management/commands/cleanup_jobs.py [hunk 10]
```diff
@@ -21,7 +23,160 @@
...
+            self.parts_to_drop.add(old_events_tbl_name)
+
+    def drop_partitions(self):
+        if self.parts_to_drop and not self.dry_run:
+            parts_to_drop_str = ','.join(self.parts_to_drop)
+            self.logger.debug(f"Dropping {parts_to_drop_str}")
+            with connection.cursor() as cursor:
+                cursor.execute(
+                    f"DROP TABLE {parts_to_drop_str}"
+                )
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:171]:
```
else:
    <log what we would do in dry run?>
```

File: awx/main/management/commands/cleanup_jobs.py [hunk 11]
```diff
@@ -21,7 +23,160 @@
...
+
+    def delete(self):
```
Comment (jladdjr) [inline awx/main/management/commands/cleanup_jobs.py:173]: I really dig how each operation was split up into each of the helper methods. I found this really easy to follow 💯

File: awx/main/management/commands/cleanup_jobs.py (hunk 2 above)
Comment (ryanpetrello) [reply awx/main/management/commands/cleanup_jobs.py:65]: We should definitely make sure all of our logic everywhere is assuming UTC time.

File: awx/main/management/commands/cleanup_jobs.py [hunk 12]
```diff
@@ -21,7 +23,160 @@
...
+
+        self.parts_no_drop = set([k for k, v in part_drop.items() if v is False])
+
+    def delete_jobs(self):
+        if not self.dry_run:
+            self.job_class.objects.filter(pk__in=self.jobs_pk_list).delete()
+
+    def find_partitions_to_drop(self):
+        tbl_name = unified_job_class_to_event_table_name(self.job_class)
+        old_events_tbl_name = f"{tbl_name}_old_events"
```
Comment (ryanpetrello) [inline awx/main/management/commands/cleanup_jobs.py:126]: If we renamed the old_events table to just be `{tbl_name}_{datetime_migration_ran}`, in the same `YYYYMMDD_HH` format, we wouldn't need any special logic to handle `_old_events`.

**Comment by jladdjr** [comment]
We used to have logic that would delete parent objects:
https://github.com/jladdjr/awx/pull/18/files#diff-cc408dfef8849d2b50cee4a861692b3dcc749222ae4eae896777cdcc0fe1576cL114-L123

With the current approach it seems like the following scenario is possible:
1. User kicks off a long-running job, that will trigger other jobs (inv update, project sync, or perhaps a workflow node's job, etc.). I'll just call those the child jobs.
2. Parent job starts during hour X.
3. Child job starts and completes during hour X + 1. 
4. Parent job continues running after this though, perhaps spawning other child jobs.
5. User launches a clean-up job. Partition X is ignored because the parent job is still running. Partition X + 1 is cleaned up because all child jobs completed then.

Result:
Parent job will have dangling references to child jobs?

This scenario feels a bit contrived, but still possible. It requires a parent job that launches in one hour, child jobs that launch and complete during a subsequent hour, and, finally, for the parent job to continue to have work after that hour block. I'm not sure how often customers have jobs that run for several hours, but this may be more frequent than I suspect. I'm not seeing any handling for this scenario in the new clean-up code. 

When I was starting to poke around on this I added a migration that dropped cascade deletes. Maybe we do want cascade deletes to prevent dangling references?
https://github.com/jladdjr/awx/commit/30ea17bb1d699bee304a397494b931fb4ec32919

File: awx/main/management/commands/cleanup_jobs.py (hunk 12 above)
Comment (jladdjr) [reply awx/main/management/commands/cleanup_jobs.py:126]: That's a good point.

File: awx/main/management/commands/cleanup_jobs.py (hunk 2 above)
Comment (jladdjr) [reply awx/main/management/commands/cleanup_jobs.py:65]: Good to know 👍

**Review by jladdjr — APPROVED** [review]

**Comment by chrismeyersfsu** [comment]
We want cascade NULL. In Django cascade delete is handled by Django and NOT by Postgres, this sucks. I think the main reason Django does this is so that it can trigger any signals (python code) that is registered. It's terribly inperformant and is the reason seth had to add deletion optimization to the cleanup code.

It's fine to have a dangling reference between, for example, project update and a Job. We already handle that outside of what this feature touches.

This code handles cleaning up deleted jobs job events with the set_null constraint set in postgres.

**Comment by jladdjr** [comment]
@chrismeyersfsu - I'm happy to go ahead and merge this unless you have anything else you're wanting to tinker on here

**Comment by jladdjr** [comment]
We're holding off on merging until we settle on a migration strategy

**Comment by jladdjr** [comment]
Closing out since our work has moved elsewhere

**Pull request closed without merging**